OpenRA.Mods.Common/Graphics/ActorPreview.cs

```diff
@@ -67,20 +67,35 @@ public Func<WRot> GetOrientation()
 			if (facingInfo == null)
 				return () => WRot.None;
 
+			WAngle? facing = null;
+
 			// Dynamic facing takes priority
 			var dynamicInit = reference.GetOrDefault<DynamicFacingInit>();
 			if (dynamicInit != null)
 			{
-				// TODO: Account for terrain slope
 				var getFacing = dynamicInit.Value;
-				return () => WRot.FromYaw(getFacing());
+				facing = getFacing();
 			}
 
-			// Fall back to initial actor facing if an Init isn't available
-			var facingInit = reference.GetOrDefault<FacingInit>();
-			var facing = facingInit != null ? facingInit.Value : facingInfo.GetInitialFacing();
-			var orientation = WRot.FromYaw(facing);
-			return () => orientation;
+			if (facing == null)
+			{
+				// Fall back to initial actor facing if an Init isn't available
+				var facingInit = reference.GetOrDefault<FacingInit>();
+				facing = facingInit != null ? facingInit.Value : facingInfo.GetInitialFacing();
+			}
+
+			var mobileInfo = Actor.TraitInfoOrDefault<MobileInfo>();
+			var location = reference.GetOrDefault<LocationInit>();
+			if (location == null || mobileInfo == null || mobileInfo.TerrainOrientationAdjustmentMargin.Length < 0)
+			{
+				return () => WRot.FromYaw((WAngle)facing);
+			}
+
+			var map = World.Map;
+			var terrainOrientation = map.TerrainOrientation(location.Value);
+			var terrainAdjustedFacing = new WRot(new WVec(0, 0, 1024).Rotate(terrainOrientation), (WAngle)facing);
```

Comment:
```suggestion
			var terrainAdjustedFacing = new WRot(new WVec(0, 0, 1024).Rotate(terrainOrientation), facing.Value);
```

This is the proper way of removing the nullable flag. However with the first suggestion we make sure it can't be null either way so you can remove this completely